<!DOCTYPE html>
<html lang="en">

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Introduction to BioPython</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="../assets/main.css">
  <link rel="canonical" href="http://localhost:4000/jekyll/update/2017/07/07/biopython.html">
  <link rel="alternate" type="application/rss+xml" title="RITct Blog" href="/feed.xml">


</head>


  <body>

    <header class="site-header" role="banner">

  <div class="wrapper">


    <a class="site-title" href="../blog.html">RITct Blog</a>


      <nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
              <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
              <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger">
            <a href="../../index.html">Home</a> | <a href="../about.html">About</a>
        </div>
      </nav>

  </div>
</header>


    <main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title" itemprop="name headline">Introduction to BioPython</h1>
    <p class="post-meta">
      <time datetime="2017-07-07T13:32:02+05:30" itemprop="datePublished">

        Akshatha Nayak
      </time>
      </p>
  </header>

  <div class="post-content" itemprop="articleBody">
    <p><img src="../../assets/images/biopython.jpeg" alt="biopython" /></p>

<p><strong>WHAT’S BIOPYTHON</strong></p>

<p>Biopython is a library containing freely available Python tools for computational Biology. It makes it easy to write python programs for bioinformatics use.</p>

<p>The basic functionalities provided by Biopython include :</p>

<ul>
  <li>Parsers for various Bioinformatics file formats (BLAST output, FASTA, Genbank etc)</li>
  <li>Access to online services (NCBI, Expasy etc)</li>
  <li>A standard sequence class that deals with sequences, ids on sequences, and sequence features.</li>
  <li>Interface to some common Bioinformatics programs like BLAST from NCBI (tool for sequence alignment) and EMBOSS tools (tools for sequence analysis)</li>
  <li>Tools for performing common operations on sequences, such as translation, transcription and weight calculations</li>
  <li>Integration with BioSQL, a sequence database schema</li>
</ul>

<p><strong>INSTALLATION</strong></p>

<p>Before installing BioPython, you need to install the prerequisites, i.e python and NumPy.</p>

<ul>
  <li>Python</li>
</ul>

<p>Biopython is available for python 2.6, 2.7, 3.3, 3.4, 3.5 (<a href="https://www.python.org/">http://www.python.org</a>).</p>

<ul>
  <li>NumPy (Numerical Python )</li>
</ul>

<p>To install NumPy , you can use pip :</p>
<div class="language-bash highlighter-rouge"><pre class="highlight"><code>    pip install numpy
</code></pre>
</div>
<ul>
  <li>Biopython</li>
</ul>

<p>After installing numpy, you can install biopython using pip:</p>
<div class="language-bash highlighter-rouge"><pre class="highlight"><code>    pip install biopython
</code></pre>
</div>
<p>To check if biopython is install properly, use this:</p>
<div class="highlighter-rouge"><pre class="highlight"><code>    import Bio
</code></pre>
</div>
<p>If this gives an error, Biopython is not installed.</p>

<p>For detailed instructions on installation, refer <a href="https://pypi.python.org/pypi/biopython/1.67">Python Package Index</a>.</p>

<p><strong>GETTING STARTED</strong></p>

<p>Let’s look at some of the functionalities that make Biopython awesome !</p>

<ul>
  <li>Parsing various Biological file formats</li>
</ul>

<p>Most of the biological data is stored as special file formats such as FASTA formatted text files, GENBANK formatted text files etc. Parsing these file formats into a format that can be manipulated using a programming language is a challenging task that can be simplified using the parsers provided in biopython.</p>

<p>Suppose you have a file named “ABC” in FASTA format. You can parse the file format and obtain a list of sequnces stored in the file as follows:</p>

<div class="language-python highlighter-rouge"><pre class="highlight"><code>    <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
    <span class="n">records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s">"ABC.fasta"</span><span class="p">,</span> <span class="s">"fasta"</span><span class="p">))</span>
</code></pre>
</div>

<p>A file in GENBANK format can be parsed similarly</p>

<div class="language-python highlighter-rouge"><pre class="highlight"><code>    <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
    <span class="n">records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s">"ABC.gbk"</span><span class="p">,</span> <span class="s">"genbank"</span><span class="p">))</span>
</code></pre>
</div>

<ul>
  <li>Accessing Online Databases</li>
</ul>

<p>Biopython can be used to access and download biological data from several databases such as NCBI, Expasy etc. The Bio.Entrez module can be used for this purpose.</p>

<div class="language-python highlighter-rouge"><pre class="highlight"><code>    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

    <span class="c">#provide your email-id</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">"me@email.com"</span>

    <span class="c">#IDs to be searched</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">records</span> <span class="o">=</span> <span class="p">[</span><span class="s">"P68871"</span><span class="p">,</span><span class="s">"Q96I25"</span><span class="p">]</span>

    <span class="c">#search the database and obtain data in FASTA format</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="k">for</span> <span class="n">rec_id</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
    <span class="o">...</span>     <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">"protein"</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">rec_id</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">"fasta"</span><span class="p">)</span>
    <span class="o">...</span>     <span class="n">seqRec</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span><span class="s">"fasta"</span><span class="p">)</span>
    <span class="o">...</span>     <span class="k">print</span><span class="p">(</span><span class="n">seqRec</span><span class="p">)</span>
    <span class="o">...</span>     <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="o">...</span>

    <span class="n">ID</span><span class="p">:</span> <span class="n">P68871</span><span class="o">.</span><span class="mi">2</span>
    <span class="n">Name</span><span class="p">:</span> <span class="n">P68871</span><span class="o">.</span><span class="mi">2</span>
    <span class="n">Description</span><span class="p">:</span> <span class="n">P68871</span><span class="o">.</span><span class="mi">2</span> <span class="n">RecName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">Hemoglobin</span> <span class="n">subunit</span> <span class="n">beta</span><span class="p">;</span> <span class="n">AltName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">Beta</span><span class="o">-</span><span class="n">globin</span><span class="p">;</span> <span class="n">AltName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">Hemoglobin</span> <span class="n">beta</span> <span class="n">chain</span><span class="p">;</span> <span class="n">Contains</span><span class="p">:</span> <span class="n">RecName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">LVV</span><span class="o">-</span><span class="n">hemorphin</span><span class="o">-</span><span class="mi">7</span><span class="p">;</span> <span class="n">Contains</span><span class="p">:</span> <span class="n">RecName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">Spinorphin</span>
    <span class="n">Number</span> <span class="n">of</span> <span class="n">features</span><span class="p">:</span> <span class="mi">0</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDA...KYH'</span><span class="p">,</span> <span class="n">SingleLetterAlphabet</span><span class="p">())</span>

    <span class="n">ID</span><span class="p">:</span> <span class="n">Q96I25</span><span class="o">.</span><span class="mi">1</span>
    <span class="n">Name</span><span class="p">:</span> <span class="n">Q96I25</span><span class="o">.</span><span class="mi">1</span>
    <span class="n">Description</span><span class="p">:</span> <span class="n">Q96I25</span><span class="o">.</span><span class="mi">1</span> <span class="n">RecName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">Splicing</span> <span class="n">factor</span> <span class="mi">45</span><span class="p">;</span> <span class="n">AltName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="mi">45</span> <span class="n">kDa</span><span class="o">-</span><span class="n">splicing</span> <span class="n">factor</span><span class="p">;</span> <span class="n">AltName</span><span class="p">:</span> <span class="n">Full</span><span class="o">=</span><span class="n">RNA</span><span class="o">-</span><span class="n">binding</span> <span class="n">motif</span> <span class="n">protein</span> <span class="mi">17</span>
    <span class="n">Number</span> <span class="n">of</span> <span class="n">features</span><span class="p">:</span> <span class="mi">0</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'MSLYDDLGVETSDSKTEGWSKNFKLLQSQLQVKKAALTQAKSQRTKQSTVLAPV...EQV'</span><span class="p">,</span> <span class="n">SingleLetterAlphabet</span><span class="p">())</span>
</code></pre>
</div>
<ul>
  <li>Sequence Objects and common operations</li>
</ul>

<p>Biopython has sequence objects that are basically strings of letters. We can perform operations such as indexing, calculating string length, iterating through the characters , slicing the string etc, just like we do with python strings.</p>

<div class="language-python highlighter-rouge"><pre class="highlight"><code>    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>

    <span class="o">&gt;&gt;&gt;</span> <span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s">"ATGCGTACGATACATACAGCGT"</span> <span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>

    <span class="o">&gt;&gt;&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>		<span class="c">#length of sequence</span>
    <span class="mi">22</span>

    <span class="o">&gt;&gt;&gt;</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">"A"</span><span class="p">)</span>	<span class="c">#count occurrences of a character</span>
    <span class="mi">7</span>

    <span class="o">&gt;&gt;&gt;</span> <span class="n">my_seq</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">7</span><span class="p">]</span>		<span class="c">#slicing the sequence</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'CGTA'</span><span class="p">,</span> <span class="n">IUPACUnambiguousDNA</span><span class="p">())</span>

    <span class="o">&gt;&gt;&gt;</span> <span class="k">for</span> <span class="n">letter</span> <span class="ow">in</span> <span class="n">my_seq</span><span class="p">[:</span><span class="mi">5</span><span class="p">]:</span>
    <span class="o">...</span>     <span class="k">print</span><span class="p">(</span><span class="n">letter</span><span class="p">)</span>	<span class="c">#iterating through characters </span>
    <span class="o">...</span>
    <span class="n">A</span>
    <span class="n">T</span>
    <span class="n">G</span>
    <span class="n">C</span>
    <span class="n">G</span>
</code></pre>
</div>

<p>Though the <code class="highlighter-rouge">Seq</code> objects in Biopython and standard Python strings have some similarities, there are two major differences.</p>

<ul>
  <li>The <code class="highlighter-rouge">Seq</code> object supports many biologically relevant methods like <code class="highlighter-rouge">translate()</code>, <code class="highlighter-rouge">transcribe()</code>, <code class="highlighter-rouge">reverse_complement()</code> etc.</li>
  <li>The <code class="highlighter-rouge">Seq</code> object has an important attribute, <code class="highlighter-rouge">alphabet</code>, which is an object describing what the individual characters making up the sequence string mean, and how they should be interpreted. For example, a <code class="highlighter-rouge">Seq</code> object could denote a protein sequence, or a DNA sequence.</li>
</ul>

<p>Some of the biologically relevant operations that can be performed using Biopython methods are demonstrated below.</p>

<div class="language-python highlighter-rouge"><pre class="highlight"><code>    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">GC</span>

    <span class="o">&gt;&gt;&gt;</span> <span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s">"ATGCGTACGATACATACAGCGT"</span> <span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>

    <span class="c">#calculating GC content of the DNA sequence</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">GC</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
    <span class="mf">45.45454545454545</span>

    <span class="c">#complement of DNA sequence</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'TACGCATGCTATGTATGTCGCA'</span><span class="p">,</span> <span class="n">IUPACUnambiguousDNA</span><span class="p">())</span>

    <span class="c">#reverse complement of DNA</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'ACGCTGTATGTATCGTACGCAT'</span><span class="p">,</span> <span class="n">IUPACUnambiguousDNA</span><span class="p">())</span>

    <span class="c">#simulating biological DNA strands</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">coding_dna</span> <span class="o">=</span> <span class="n">my_seq</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">template_dna</span> <span class="o">=</span> <span class="n">coding_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">template_dna</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'ACGCTGTATGTATCGTACGCAT'</span><span class="p">,</span> <span class="n">IUPACUnambiguousDNA</span><span class="p">())</span>

    <span class="c">#transcription process (DNA -&gt; mRNA)</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">messenger_rna</span> <span class="o">=</span> <span class="n">template_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">messenger_rna</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'AUGCGUACGAUACAUACAGCGU'</span><span class="p">,</span> <span class="n">IUPACUnambiguousRNA</span><span class="p">())</span>

    <span class="c">#translation process (mRNA -&gt; Protein)</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">protein</span> <span class="o">=</span> <span class="n">messenger_rna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
    <span class="o">&gt;&gt;&gt;</span> <span class="n">protein</span>
    <span class="n">Seq</span><span class="p">(</span><span class="s">'MRTIHTA'</span><span class="p">,</span> <span class="n">IUPACProtein</span><span class="p">())</span>
</code></pre>
</div>

<p>Apart from these, Biopython offers lots of other features. So, if you are interested in bioinformatics, and love to program in python, then Biopython is the perfect choice for you !</p>

<p>To know more, check out the official <a href="http://biopython.org/wiki/Documentation">documentation</a>.</p>

<p><strong>USEFUL RESOURCES</strong></p>

<ul>
  <li>The <a href="http:// http://biopython.org/DIST/docs/tutorial/Tutorial.html">Biopython Tutorial and Cookbook</a> contains the official Biopython documentation and tutorials to get you started.</li>
  <li><a href="http://www1.chapman.edu/~fahy/bioinformatics/">Guide to Bioinformatics with Biopython</a></li>
  <li><a href="https://krother.gitbooks.io/biopython-tutorial/content/">Biopython Tutorial by Dr. Kristian Rother</a></li>
</ul>

  </div>


</article>

      </div>
    </main>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">RITct Blog</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>

              RITct Blog

            </li>

            <li><a href="mailto:admin@ritcreativeteam.ml">admin@ritcreativeteam.ml</a></li>

        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">

          <li>
            <a href="https://github.com/RITct"><span class="icon icon--github"><svg viewBox="0 0 16 16" width="16px" height="16px"><path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/></svg>
</span><span class="username">RITct</span></a>

          </li>
          <li>
          <a href="https://facebook.com/TinkerHubRIT" target="_none"><span class="icon icon--facebook"><svg viewBox="0 0 16 16" width="16px" height="16px"><path fill="#828282" d="M15.117 0H.883C.395 0 0 .395 0 .883v14.234c0 .488.395.883.883.883h7.663V9.804H6.46V7.39h2.086V5.607c0-2.066 1.262-3.19 3.106-3.19.883 0 1.642.064 1.863.094v2.16h-1.28c-1 0-1.195.48-1.195 1.18v1.54h2.39l-.31 2.42h-2.08V16h4.077c.488 0 .883-.395.883-.883V.883C16 .395 15.605 0 15.117 0"></svg>
          </span><span class="username">TinkerHub RIT</span></a>
          </li>


        </ul>
      </div>

      <div class="footer-col footer-col-3">
        <p>Tutorials, hacks, experiences and much more from the RIT Creative Team
</p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
